cardano-api/src/Cardano/Api/Typed.hs

```diff
@@ -1565,6 +1561,110 @@ data MultiSigScript = RequireSignature (Hash PaymentKey)
+                                    ]
+                                  ]
+
+  | otherwise = error $ "Cardano.Api.Typed.toJSONmofnChecks failure occured: " ++ " required: "
+                      ++ show required ++ " total: " ++  show total
+                      ++ " number of key hashes: " ++ show (length keys)
+
+instance FromJSON MultiSigScript where
+  parseJSON = Aeson.withObject "MultiSigScript" $ \obj ->
+   select [all' obj, any' obj, mofn (Object obj)]
+
+select :: [ParseMultiSigScript] -> Aeson.Parser MultiSigScript
+select [] = fail "No multisig scripts found"
+select (x : xs) = case x of
+                    All mss -> mss
+                    Any mss -> mss
+                    MofN mss  -> mss
+                    ScriptNotFound -> select xs
+
+data ParseMultiSigScript = All (Aeson.Parser MultiSigScript)
+                         | Any (Aeson.Parser MultiSigScript)
+                         | MofN (Aeson.Parser MultiSigScript)
+                         | ScriptNotFound
+
+-- Parse "all" multisig objects
+all' :: HMS.HashMap Text Value -> ParseMultiSigScript
+all' obj = case HMS.lookup "all" obj of
+             Just (Array vecAllReqHashes) ->
+               All . return . RequireAllOf $ gatherSignatures vecAllReqHashes
+             _ -> ScriptNotFound
+
+-- Parse "any" multisig objects
+any' :: HMS.HashMap Text Value -> ParseMultiSigScript
+any' obj = case HMS.lookup "any" obj of
+             Just (Array vecAnyReqHashes) ->
+               Any . return . RequireAnyOf $ gatherSignatures vecAnyReqHashes
+             _ -> ScriptNotFound
+
+-- Parse "mofn" multisig objects
+mofn :: Value -> ParseMultiSigScript
+mofn val =
+  MofN $ Aeson.withObject "mofn"
+           (\o -> do mofnObj <- o .: "atLeast"
+                     required <- mofnObj .: "required"
+                     total <- mofnObj .: "total"
+                     keyHashes <- case HMS.lookup "paymentKeyHashes" mofnObj of
+                                    Just (Array keyhashes) -> return $ gatherSignatures keyhashes
+                                    _ -> return []
+                     fromJSONmOfnChecks keyHashes required total)
+           val
+
+
+
+fromJSONmOfnChecks :: [MultiSigScript] -> Int -> Int -> Aeson.Parser MultiSigScript
+fromJSONmOfnChecks keys required total
+  | required == 1 = fail $ "required is equal to one, you should use the \"any\" multisig script"
+  | required == total = fail "required = total, you should use the \"all\" multisig script"
+  | length keys < required = fail $ "required exceeds the number of payment key hashes. \
+                                    \Number of keys: " ++ show (length keys)
+                                  ++ " required: " ++ show required
+  | length keys == total = return $ RequireMOf required keys
+  | otherwise = fail $ "Cardano.Api.Typed.fromJSONmOfnChecks failure occured. required: "
+                     ++ show required ++ " total: " ++ show total
+                     ++ " keys: " ++ show keys
```

Comment:
Should we also check whether `required <= 0` as well as whether `required > total`?

Reply:
I think we should. I'll add in a check for that.